### ISSUE TITLE
Workflow using resource quotas for cluster level extended resources

### ISSUE THREAD

**Issue by tse77** [argoproj/argo#4170]
I'm trying to get Argo to work with resource quotas for extended resources but I cant get it to work.

I have my own custom scheduler where I defined my managed resources, I included a dummy scheduler with one managed resource (example.com/foo):

``` yml
apiVersion: v1
kind: ServiceAccount
metadata:
  name: dummy-scheduler
  namespace: kube-system
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: dummy-scheduler-cluster-admin
  namespace: kube-system
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: cluster-admin
subjects:
  - kind: ServiceAccount
    namespace: kube-system
    name: dummy-scheduler
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: dummy-scheduler-config
  namespace: kube-system
data:
  config.yaml: |
    apiVersion: kubescheduler.config.k8s.io/v1alpha1
    kind: KubeSchedulerConfiguration
    schedulerName: dummy-scheduler
    algorithmSource:
      policy:
        configMap:
          namespace: kube-system
          name: dummy-scheduler-policy
    leaderElection:
      leaderElect: true
      lockObjectName: dummy-scheduler
      lockObjectNamespace: kube-system
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: dummy-scheduler-policy
  namespace: kube-system
data:
 policy.cfg : |
  {
    "kind" : "Policy",
    "apiVersion" : "v1",
    "predicates" : [
      {"name" : "PodFitsHostPorts"},
      {"name" : "PodFitsResources"},
      {"name" : "NoDiskConflict"},
      {"name" : "MatchNodeSelector"},
      {"name" : "HostName"}
    ],
    "priorities" : [
      {"name" : "LeastRequestedPriority", "weight" : 1},
      {"name" : "BalancedResourceAllocation", "weight" : 1},
      {"name" : "ServiceSpreadingPriority", "weight" : 1},
      {"name" : "EqualPriority", "weight" : 1}
    ],
    "extenders" : [{
      "urlPrefix": "",
      "filterVerb": "",
      "prioritizeVerb": "",
      "preemptVerb": "",
      "bindVerb": "",
      "weight": 1,
      "enableHttps": false,
      "nodeCacheCapable": false,
      "managedResources" : [
        {
          "name": "example.com/foo",
          "ignoredByScheduler": true
        }]
    }],
    "hardPodAffinitySymmetricWeight" : 10
  }
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: dummy-scheduler
  namespace: kube-system
  labels:
    app: dummy-scheduler
spec:
  replicas: 1
  selector:
    matchLabels:
      app: dummy-scheduler
  template:
    metadata:
      labels:
        app: dummy-scheduler
    spec:
      serviceAccountName: dummy-scheduler
      volumes:
      - name: dummy-scheduler-config
        configMap:
          name: dummy-scheduler-config
      containers:
      - name: dummy-scheduler-ctr
        image: gcr.io/google_containers/hyperkube:v1.17.0
        imagePullPolicy: IfNotPresent
        args:
        - kube-scheduler
        - --config=/dummy-scheduler/config.yaml
        - -v=4
        volumeMounts:
        - name: dummy-scheduler-config
          mountPath: /dummy-scheduler
```

I also have created resource quotas like so:

``` yaml
apiVersion: v1
kind: ResourceQuota
metadata:
  name: my-extended-resources
  namespace: argo
spec:
  hard:
    requests.example.com/foo: 1
    limits.example.com/foo: 1
```

When I run this simple workflow:

``` yml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: counter-
spec:
  entrypoint: counter
  templates:
  - name: counter
    schedulerName: dummy-scheduler
    container:
      image: alpine
      command: ["/bin/sh"]
      args: ["-c", "n=1; while [ $n -lt 30 ]; do echo $n; sleep 1; n=$((n+1));done"]
      resources:
        requests:
          "example.com/foo": 1
        limits:
          memory: 32Mi
          cpu: 100m
          "example.com/foo": 1
```

I get this error when submitting the work:
``` shell
STEP              TEMPLATE  PODNAME        DURATION  MESSAGE
 ● counter-k2wl7  counter   counter-k2wl7  1s        pods "counter-k2wl7" is forbidden: exceeded quota: my-extended-resources, requested: requests.example.com/foo=1, used: requests.example.com/foo=1, limited: requests.example.com/foo=1
```
 
But here is no other object that have requested the resource except this workflow. However the workflow continues to run until finished and when I look at my-extended-resource I can see that it is used:
 
``` shell
 k -n argo describe resourcequotas my-extended-resources
Name:                     my-extended-resources
Namespace:                argo
Resource                  Used  Hard
--------                  ----  ----
limits.example.com/foo    0     1
requests.example.com/foo  1     1
```

If I submit more workflows that use this resource, while one workflow is executing and using the resource I get the same error message but those jobs are canceled and will not be automatically started when the resource is released as I thought.

When I for example runs a K8s Job with the same resources defined everything works ok. If the resource is used the jobs will be put back in the scheduler queue and will eventually be scheduled when the resource is released again.

Example K8s job that works:

``` yml
apiVersion: batch/v1
kind: Job
metadata:
  generateName: job-counter-
  namespace: argo
spec:
  template:
    spec:
      schedulerName: dummy-scheduler
      containers:
      - name: counter
        image: alpine
        command: ["/bin/sh"]
        args: ["-c", "n=1; while [ $n -lt 30 ]; do echo $n; sleep 1; n=$((n+1));done" ]
        resources:
          requests:
            example.com/foo: 1
          limits:
            example.com/foo: 1
      restartPolicy: Never
  backoffLimit: 1
```

How is it supposed to work in Argo?

Regards
Thomas

**Comment by tse77** [comment]
I could get it to work by setting `resubmitPendingPods: true` so it works it is just the workflows will execute while it's state is reported as pending but it's actually executing. So it will go from pending to finished. I can look at the log file and see it is updated but as I just mentioned the workflow is in pending state.

**Comment by alexec** [comment]
Please include version. This is unlikely to be a workflow issue, as it is related to resources which is core Kubernetes feature. Can you please upgrade to v2.11.1.

**Comment by tse77** [comment]
I had v2.10.2 version but upgraded to v2.11.1 and the bug in the web ui seems to be fixed now when I clicked on a specific workflow it's state now show's correctly. But there seems to be a related bug from the overview of all workflows as well. If I submit several workflow using the same resource one will be executing and the rest in pending state. But in the overview they are all shown as running and if I filter on pending by clicking in the left phases sidebar no pending workflows are shown but those are shown in the running phase instead.

Except for that the functionality seems to work as expected now.

**Comment by alexec** [comment]
Thank you @tse77 could I ask you to raise another issue for you other problem and include screenshots?

**Comment by sarabala1979** [comment]
> I could get it to work by setting `resubmitPendingPods: true` so it works it is just the workflows will execute while it's state is reported as pending but it's actually executing. So it will go from pending to finished. I can look at the log file and see it is updated but as I just mentioned the workflow is in pending state.

can you file the issue for a state didn't change with reproducible examples?

**Comment by tse77** [comment]
Yes I can do that within some days if needed. I just need to get some more experience of Argo and see how it is supposed to work by design and perhaps get back to you if I perhaps have some more questions etc. Closing this now and thanks for the help again =)
